rbac/migration_tool/in_memory_tuples.py

```diff
@@ -25,15 +25,202 @@ class RelationTuple(NamedTuple):
 
     def stringify(self):
         """Display all attributes in one line."""
+        subject_part = f"{self.subject_type_namespace}/{self.subject_type_name}:{self.subject_id}"
+        if self.subject_relation:
+            subject_part += f"#{self.subject_relation}"
         return (
             f"{self.resource_type_namespace}/{self.resource_type_name}:{self.resource_id}#{self.relation}"
-            f"@{self.subject_type_namespace}/{self.subject_type_name}:{self.subject_id}"
+            f"@{subject_part}"
         )
 
 
+RelationPredicate = Callable[["RelationTuple"], bool]
 T = TypeVar("T", bound=Hashable)
 
 
+class TupleSet:
+    """A set of relation tuples with various utility methods."""
+
+    def __init__(self, full_set: "InMemoryTuples", filtered_set: Set[RelationTuple]):
+        """Initialize the store."""
+        self._full_set = full_set
+        self._set = filtered_set
+
+    def __len__(self):
+        """Return the number of tuples in the list."""
+        return len(self._set)
+
+    def __iter__(self):
+        """Return an iterator over the tuples."""
+        return iter(self._set)
+
+    def __contains__(self, item: RelationTuple):
+        """Check if a tuple is in the list."""
+        return item in self._set
+
+    def __str__(self):
+        """Return a string representation of the store."""
+        return str(self._set)
+
+    def __repr__(self):
+        """Return a representation of the store."""
+        return f"TupleSet({repr(self._set)})"
+
+    def count_tuples(self, predicate: RelationPredicate = lambda _: True) -> int:
+        """Count tuples matching the given predicate."""
+        return len(self.find_tuples(predicate))
+
+    def find_tuples(self, predicate: RelationPredicate = lambda _: True) -> "TupleSet":
+        """Find tuples matching the given predicate."""
+        return TupleSet(self._full_set, {rel for rel in self._set if predicate(rel)})
+
+    def find_tuples_grouped(
+        self, predicate: RelationPredicate, group_by: Callable[[RelationTuple], T]
+    ) -> dict[T, "TupleSet"]:
+        """Filter tuples and group them by a key."""
+        grouped_tuples: dict[T, set[RelationTuple]] = defaultdict(set)
+        for rel in self._set:
+            if predicate(rel):
+                key = group_by(rel)
+                grouped_tuples[key].add(rel)
+        return {key: TupleSet(self._full_set, value) for key, value in grouped_tuples.items()}
+
+    def find_group_with_tuples(
+        self,
+        predicates: List[RelationPredicate],
+        group_by: Callable[[RelationTuple], T],
+        group_filter: Callable[[T], bool] = lambda _: True,
+        require_full_match: bool = False,
+        match_once: bool = True,
+    ) -> Tuple[dict[T, "TupleSet"], dict[T, "TupleSet"]]:
```

Comment:
These methods are just moved from InMemoryTuples.

Basically we reorganized things a little bit so that results from InMemoryTuples return a TupleSet type which can be further filtered. The novelty of this PR is that it adds a `traverse_subjects` method in this queryset which is not a filter but a traversal.